src/library/scala/collection/mutable/Set.scala

```diff
@@ -78,8 +78,18 @@ trait SetOps[A, +CC[X], +C <: SetOps[A, CC, C]]
     *             are removed.
     */
   def filterInPlace(p: A => Boolean): this.type = {
-    for (elem <- this.toList) // scala/bug#7269 toList avoids ConcurrentModificationException
-      if (!p(elem)) this -= elem
+    if (nonEmpty) {
+      val array = this.toArray[Any] // scala/bug#7269 toArray avoids ConcurrentModificationException
```

Comment:
Is it better to build the list of filtered elements here? 
Assuming `O(n)` for the `.toArray` call
It would reduce the second loops time complexity to `O(m)`  where m is the number of filtered elements. A special optimisation could be made when all elements are being filtered to call `clear`. And if none are being filtered to do nothing. 

Just a thought. It might not actually make a difference 😂

Reply:
What about if an array was created to the same size as the `Set`.
```
val filtered = new Array[A](this.size())
var index = 0
foreach { a =>
 // this would be another while loop
 if(!p(a)) {
   filter(index) = a
   index += 1 
 }
}
if(index > 0) {
  // clear all filter elements
}
```

Reply:
That's computationally equivalent, but more complicated, and possibly with slightly worse memory locality, than the existing algorithm.

Reply:
Ah ok fair enough. Just pondering really 😅. Cheers